File: Python/14-Backtracking/Algorithms/3-HamiltonianCycle/HamiltonianCycle.py
"""
Hamiltonian Cycle Problem:

Given a graph, determine if a Hamiltonian cycle exists. A Hamiltonian cycle is a
path that visits every vertex exactly once and returns to the starting vertex.

This implementation provides five different approaches to solve the Hamiltonian Cycle
problem using backtracking:

1.  **Basic Backtracking:** A straightforward recursive backtracking approach.
2.  **Backtracking with Adjacency Matrix:** Uses an adjacency matrix
    representation of the graph.
3.  **Backtracking with Early Pruning:** Optimizes the basic backtracking
    approach by adding a condition to check if a node is a dead end.
4.  **Backtracking with Degree Check:** Adds a degree check for each node
    before starting the backtracking to quickly rule out impossible graphs.
5. **Backtracking with Iterative Approach:** Uses a stack-based iterative approach.

Each approach includes detailed comments to explain the logic and workings of the algorithm.
"""

def is_safe(v, graph, path, pos):
    """
    Checks if vertex v can be added to the current path.

    Args:
        v: The vertex to check.
        graph: The graph represented as an adjacency list.
        path: The current path being constructed.
        pos: The current position in the path.

    Returns:
        True if the vertex can be added, False otherwise.
    """
    # Check if this vertex is an adjacent vertex of the previously added vertex.
    if graph[path[pos - 1]][v] == 0:
        return False

    # Check if the vertex has already been included in the path.  A Hamiltonian
    # cycle should not have repeated vertices.
    for vertex in path:
        if vertex == v:
            return False

    return True

def hamiltonian_cycle_util(graph, path, pos, num_vertices):
    """
    Recursive utility function to find a Hamiltonian cycle using backtracking.

    Args:
        graph: The graph represented as an adjacency list.
        path: The current path being constructed.
        pos: The current position in the path.
        num_vertices: Total number of vertices in the graph.

    Returns:
        True if a Hamiltonian cycle is found, False otherwise.
    """
    # Base case: If all vertices are included in the path
    if pos == num_vertices:
        # Check if there is an edge from the last vertex to the starting vertex.
        if graph[path[pos - 1]][path[0]] == 1:
            return True
        else:
            return False

    # Try different vertices as the next vertex in the path
    for v in range(num_vertices):
        if is_safe(v, graph, path, pos):
            path[pos] = v
            # Recursively try to construct the rest of the path
            if hamiltonian_cycle_util(graph, path, pos + 1, num_vertices):
                return True
            # Backtrack: If adding vertex v doesn't lead to a solution, remove it
            path[pos] = -1

    # If no vertex can be added to the path, return False
    return False

def hamiltonian_cycle(graph):
    """
    Finds a Hamiltonian cycle in the given graph using backtracking.

    Args:
        graph: The graph represented as an adjacency list.

    Returns:
        A list representing the Hamiltonian cycle, or None if no cycle exists.
    """
    num_vertices = len(graph)
    path = [-1] * num_vertices  # Initialize the path with -1
    path[0] = 0  # Start the path from vertex 0

    if hamiltonian_cycle_util(graph, path, 1, num_vertices):
        return path
    else:
        return None

def print_cycle(cycle):
    """
    Prints the Hamiltonian cycle.

    Args:
        cycle: A list representing the Hamiltonian cycle.
    """
    if cycle is None:
        print("No Hamiltonian cycle exists.")
    else:
        print("Hamiltonian Cycle:", end=" ")
        for vertex in cycle:
            print(vertex, end=" ")
        print(cycle[0])  # Print the starting vertex again to complete the cycle

def basic_backtracking_example():
    """
    Example usage of the basic backtracking approach.
    """
    print("\n--- Basic Backtracking Example ---")
    # Example graph represented as an adjacency list
    graph = [
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
    ]
    cycle = hamiltonian_cycle(graph)
    print_cycle(cycle)
    # Example graph with no Hamiltonian Cycle
    graph2 = [
        [0, 1, 0, 0, 0],
        [1, 0, 1, 0, 0],
        [0, 1, 0, 1, 0],
        [0, 0, 1, 0, 1],
        [0, 0, 0, 1, 0],
    ]
    cycle2 = hamiltonian_cycle(graph2)
    print_cycle(cycle2)

def is_safe_matrix(v, graph, path, pos):
    """
    Checks if vertex v can be added to the current path, using an adjacency matrix.

    Args:
        v: The vertex to check.
        graph: The graph represented as an adjacency matrix.
        path: The current path being constructed.
        pos: The current position in the path.

    Returns:
        True if the vertex can be added, False otherwise.
    """
    if graph[path[pos - 1]][v] == 0:
        return False
    for vertex in path:
        if vertex == v:
            return False
    return True

def hamiltonian_cycle_util_matrix(graph, path, pos, num_vertices):
    """
    Recursive utility function to find a Hamiltonian cycle using an adjacency matrix.

    Args:
        graph: The graph represented as an adjacency matrix.
        path: The current path being constructed.
        pos: The current position in the path.
        num_vertices: The number of vertices in the graph.

    Returns:
        True if a Hamiltonian cycle is found, False otherwise.
    """
    if pos == num_vertices:
        if graph[path[pos - 1]][path[0]] == 1:
            return True
        else:
            return False

    for v in range(num_vertices):
        if is_safe_matrix(v, graph, path, pos):
            path[pos] = v
            if hamiltonian_cycle_util_matrix(graph, path, pos + 1, num_vertices):
                return True
            path[pos] = -1
    return False

def hamiltonian_cycle_matrix(graph):
    """
    Finds a Hamiltonian cycle in the given graph represented as an adjacency matrix.

    Args:
        graph: The graph represented as an adjacency matrix.

    Returns:
        A list representing the Hamiltonian cycle, or None if no cycle exists.
    """
    num_vertices = len(graph)
    path = [-1] * num_vertices
    path[0] = 0
    if hamiltonian_cycle_util_matrix(graph, path, 1, num_vertices):
        return path
    else:
        return None

def backtracking_with_adjacency_matrix_example():
    """
    Example usage of the backtracking approach with an adjacency matrix.
    """
    print("\n--- Backtracking with Adjacency Matrix Example ---")
    # Example graph represented as an adjacency matrix
    graph_matrix = [
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
    ]
    cycle_matrix = hamiltonian_cycle_matrix(graph_matrix)
    print_cycle(cycle_matrix)

    graph_matrix_2 = [
        [0, 1, 0, 0, 0],
        [1, 0, 1, 0, 0],
        [0, 1, 0, 1, 0],
        [0, 0, 1, 0, 1],
        [0, 0, 0, 1, 0],
    ]
    cycle_matrix_2 = hamiltonian_cycle_matrix(graph_matrix_2)
    print_cycle(cycle_matrix_2)

def is_safe_pruning(v, graph, path, pos):
    """
    Checks if vertex v can be added to the current path, with early pruning.

    Args:
        v: The vertex to check.
        graph: The graph represented as an adjacency list.
        path: The current path being constructed.
        pos: The current position in the path.

    Returns:
        True if the vertex can be added, False otherwise.
    """
    if graph[path[pos - 1]][v] == 0:
        return False
    for vertex in path:
        if vertex == v:
            return False
    return True

def hamiltonian_cycle_util_pruning(graph, path, pos, num_vertices):
    """
    Recursive utility function to find a Hamiltonian cycle with early pruning.

    Args:
        graph: The graph represented as an adjacency list.
        path: The current path being constructed.
        pos: The current position in the path.
        num_vertices: The number of vertices in the graph.

    Returns:
        True if a Hamiltonian cycle is found, False otherwise.
    """
    if pos == num_vertices:
        if graph[path[pos - 1]][path[0]] == 1:
            return True
        else:
            return False

    for v in range(num_vertices):
        if is_safe_pruning(v, graph, path, pos):
            path[pos] = v
            if hamiltonian_cycle_util_pruning(graph, path, pos + 1, num_vertices):
                return True
            path[pos] = -1
    return False

def hamiltonian_cycle_pruning(graph):
    """
    Finds a Hamiltonian cycle in the given graph using backtracking with early pruning.

    Args:
        graph: The graph represented as an adjacency list.

    Returns:
        A list representing the Hamiltonian cycle, or None if no cycle exists.
    """
    num_vertices = len(graph)
    path = [-1] * num_vertices
    path[0] = 0
    if hamiltonian_cycle_util_pruning(graph, path, 1, num_vertices):
        return path
    else:
        return None

def backtracking_with_early_pruning_example():
    """
    Example usage of the backtracking approach with early pruning.
    """
    print("\n--- Backtracking with Early Pruning Example ---")
    graph_pruning = [
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
    ]
    cycle_pruning = hamiltonian_cycle_pruning(graph_pruning)
    print_cycle(cycle_pruning)

    graph_pruning_2 = [
        [0, 1, 0, 0, 0],
        [1, 0, 1, 0, 0],
        [0, 1, 0, 1, 0],
        [0, 0, 1, 0, 1],
        [0, 0, 0, 1, 0],
    ]
    cycle_pruning_2 = hamiltonian_cycle_pruning(graph_pruning_2)
    print_cycle(cycle_pruning_2)

def has_valid_degree(graph):
    """
    Checks if the graph has a valid degree for a Hamiltonian Cycle.
    A Hamiltonian cycle requires that every vertex has a degree of at least 2.

    Args:
        graph: The graph represented as an adjacency list.

    Returns:
        True if every vertex has a degree of 2 or more, False otherwise.
    """
    num_vertices = len(graph)
    for i in range(num_vertices):
        degree = sum(graph[i])
        if degree < 2:
            return False
    return True

def hamiltonian_cycle_util_degree_check(graph, path, pos, num_vertices):
    """
    Recursive utility function to find a Hamiltonian cycle with a degree check.

    Args:
        graph: The graph represented as an adjacency list.
        path: The current path being constructed.
        pos: The current position in the path.
        num_vertices: The number of vertices in the graph.

    Returns:
        True if a Hamiltonian cycle is found, False otherwise.
    """
    if pos == num_vertices:
        if graph[path[pos - 1]][path[0]] == 1:
            return True
        else:
            return False

    for v in range(num_vertices):
        if is_safe_pruning(v, graph, path, pos):
            path[pos] = v
            if hamiltonian_cycle_util_degree_check(graph, path, pos + 1, num_vertices):
                return True
            path[pos] = -1
    return False

def hamiltonian_cycle_degree_check(graph):
    """
    Finds a Hamiltonian cycle in the given graph using backtracking with a degree check.

    Args:
        graph: The graph represented as an adjacency list.

    Returns:
        A list representing the Hamiltonian cycle, or None if no cycle exists.
    """
    num_vertices = len(graph)
    # Check the degree of each vertex before starting the backtracking
    if not has_valid_degree(graph):
        return None  # Return None if a vertex has degree less than 2

    path = [-1] * num_vertices
    path[0] = 0
    if hamiltonian_cycle_util_degree_check(graph, path, 1, num_vertices):
        return path
    else:
        return None
def backtracking_with_degree_check_example():
    """
    Example usage of the backtracking approach with a degree check.
    """
    print("\n--- Backtracking with Degree Check Example ---")
    graph_degree_check = [
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
    ]
    cycle_degree_check = hamiltonian_cycle_degree_check(graph_degree_check)
    print_cycle(cycle_degree_check)

    graph_degree_check_2 = [
        [0, 1, 0, 0, 0],
        [1, 0, 1, 0, 0],
        [0, 1, 0, 1, 0],
        [0, 0, 1, 0, 1],
        [0, 0, 0, 1, 0],
    ]
    cycle_degree_check_2 = hamiltonian_cycle_degree_check(graph_degree_check_2)
    print_cycle(cycle_degree_check_2)

    graph_degree_check_3 = [ # A graph where each node has degree >= 2
        [0, 1, 1, 0, 0],
        [1, 0, 1, 1, 0],
        [1, 1, 0, 0, 1],
        [0, 1, 0, 0, 1],
        [0, 0, 1, 1, 0],
    ]
    cycle_degree_check_3 = hamiltonian_cycle_degree_check(graph_degree_check_3)
    print_cycle(cycle_degree_check_3)

def hamiltonian_cycle_iterative(graph):
    """
    Finds a Hamiltonian cycle in the given graph using an iterative approach.

    Args:
        graph: The graph represented as an adjacency list.

    Returns:
        A list representing the Hamiltonian cycle, or None if no cycle exists.
    """
    num_vertices = len(graph)
    stack = [(0, [0])]  # Start from vertex 0, path starts with [0]

    while stack:
        current_vertex, path = stack.pop()
        if len(path) == num_vertices:
            if graph[current_vertex][0] == 1:
                return path
            else:
                continue  # Dead end, try other paths

        for next_vertex in range(num_vertices):
            if is_safe(next_vertex, graph, path, len(path)):
                stack.append((next_vertex, path + [next_vertex]))
    return None

def backtracking_with_iterative_approach_example():
    """
    Example usage of the backtracking approach.
    """
    print("\n--- Backtracking with Iterative Approach Example ---")
    # Example graph represented as an adjacency list
    graph_iterative = [
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
        [1, 0, 1, 0, 1],
        [0, 1, 0, 1, 0],
    ]
    cycle_iterative = hamiltonian_cycle_iterative(graph_iterative)
    print_cycle(cycle_iterative)

    graph_iterative_2 = [
        [0, 1, 0, 0, 0],
        [1, 0, 1, 0, 0],
        [0, 1, 0, 1, 0],
        [0, 0, 1, 0, 1],
        [0, 0, 0, 1, 0],
    ]
    cycle_iterative_2 = hamiltonian_cycle_iterative(graph_iterative_2)
    print_cycle(cycle_iterative_2)

if __name__ == "__main__":
    basic_backtracking_example()
    backtracking_with_adjacency_matrix_example()
    backtracking_with_early_pruning_example()
    backtracking_with_degree_check_example()
    backtracking_with_iterative_approach_example()

File: C++/14-Backtracking/Algorithms/3-HamiltonianCycle/HamiltonianCycle.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <stack> // Include the stack header

using namespace std;

// Function to check if a given vertex 'v' can be added to the current path
bool isSafe(int v, vector<int>& path, vector<vector<int>>& graph, int pos) {
    // Check if the current vertex 'v' is adjacent to the previously added vertex in the path.
    if (graph[path[pos - 1]][v] == 0)
        return false;

    // Check if the vertex 'v' is already included in the path.  A Hamiltonian cycle visits each vertex exactly once.
    for (int i = 0; i < pos; i++)
        if (path[i] == v)
            return false;

    return true;
}

// 1.  Recursive Backtracking Approach
//    - Basic backtracking implementation.
//    - Uses recursion to explore possible paths.
//    - Time Complexity: O(n!), where n is the number of vertices.
//    - Space Complexity: O(n) for the path vector and recursion stack.
bool findHamiltonianCycle_Recursive(vector<vector<int>>& graph, vector<int>& path, int pos, int start) {
    int n = graph.size();

    // Base case: All vertices have been included in the path.
    if (pos == n) {
        // Check if there is an edge from the last vertex to the starting vertex to complete the cycle.
        if (graph[path[pos - 1]][start] == 1)
            return true;
        else
            return false;
    }

    // Try different vertices as the next vertex in the path.
    for (int v = 0; v < n; v++) {
        if (isSafe(v, path, graph, pos)) {
            path[pos] = v;  // Add the vertex 'v' to the current path.
            // Recursively try to find a Hamiltonian cycle from the new path.
            if (findHamiltonianCycle_Recursive(graph, path, pos + 1, start))
                return true;
            // Backtrack: If adding 'v' doesn't lead to a solution, remove it from the path.
            path[pos] = -1;
        }
    }

    // If no vertex can be added to the path, return false.
    return false;
}

// Function to initiate the recursive backtracking approach
void solveHamiltonianCycle_Recursive(vector<vector<int>>& graph) {
    int n = graph.size();
    vector<int> path(n, -1); // Initialize the path vector with -1.
    path[0] = 0;             // Start the path from vertex 0 (you can choose any starting vertex).

    if (findHamiltonianCycle_Recursive(graph, path, 1, 0)) {
        cout << "Hamiltonian Cycle found (Recursive): ";
        for (int v : path)
            cout << v << " ";
        cout << "0" << endl; // Print the starting vertex to complete the cycle.
    } else {
        cout << "No Hamiltonian Cycle exists (Recursive)" << endl;
    }
}


// 2. Backtracking with Forward Checking
//    - Improves upon basic backtracking by adding a forward checking step.
//    - Before adding a vertex to the path, it checks if that vertex has any unvisited neighbors.
//    - This prunes the search space and can improve performance in some cases.
//    - Time Complexity: O(n!), but often faster than basic backtracking in practice.
//    - Space Complexity: O(n)
bool findHamiltonianCycle_ForwardChecking(vector<vector<int>>& graph, vector<int>& path, int pos, int start, vector<bool>& visited) {
    int n = graph.size();

    if (pos == n) {
        if (graph[path[pos - 1]][start] == 1)
            return true;
        else
            return false;
    }

    for (int v = 0; v < n; v++) {
        if (isSafe(v, path, graph, pos) && !visited[v]) {
             // Forward checking: Check if 'v' has any unvisited neighbors
            bool hasUnvisitedNeighbor = false;
            for (int neighbor = 0; neighbor < n; neighbor++) {
                if (graph[v][neighbor] == 1 && !visited[neighbor] && neighbor != start) {
                    hasUnvisitedNeighbor = true;
                    break;
                }
            }
            if (!hasUnvisitedNeighbor && pos != n-1) continue;

            path[pos] = v;
            visited[v] = true;
            if (findHamiltonianCycle_ForwardChecking(graph, path, pos + 1, start, visited))
                return true;
            path[pos] = -1;
            visited[v] = false;
        }
    }
    return false;
}

void solveHamiltonianCycle_ForwardChecking(vector<vector<int>>& graph) {
    int n = graph.size();
    vector<int> path(n, -1);
    vector<bool> visited(n, false);
    path[0] = 0;
    visited[0] = true;

    if (findHamiltonianCycle_ForwardChecking(graph, path, 1, 0, visited)) {
        cout << "Hamiltonian Cycle found (Forward Checking): ";
        for (int v : path)
            cout << v << " ";
         cout << "0" << endl;
    } else {
        cout << "No Hamiltonian Cycle exists (Forward Checking)" << endl;
    }
}



// 3. Backtracking with Degree Ordering
//    - Heuristic-guided backtracking.
//    - Tries to visit vertices with the lowest degree first.  The intuition is that these vertices are more constrained.
//    - Can significantly improve performance in some graphs.
//    - Time Complexity: Still O(n!) in the worst case, but often much faster in practice.
//    - Space Complexity: O(n)
bool findHamiltonianCycle_DegreeOrdering(vector<vector<int>>& graph, vector<int>& path, int pos, int start, vector<bool>& visited, vector<int>& degree) {
    int n = graph.size();

    if (pos == n) {
        if (graph[path[pos - 1]][start] == 1)
            return true;
        else
            return false;
    }

    // Create a vector of candidate vertices and sort them by their degree.
    vector<pair<int, int>> candidates; // (vertex, degree)
    for (int v = 0; v < n; v++) {
        if (isSafe(v, path, graph, pos) && !visited[v]) {
            candidates.push_back({v, degree[v]});
        }
    }
    sort(candidates.begin(), candidates.end(), [](const auto& a, const auto& b) {
        return a.second < b.second; // Sort in ascending order of degree
    });

    for (const auto& candidate : candidates) {
        int v = candidate.first;
        path[pos] = v;
        visited[v] = true;
        if (findHamiltonianCycle_DegreeOrdering(graph, path, pos + 1, start, visited, degree))
            return true;
        path[pos] = -1;
        visited[v] = false;
    }
    return false;
}

void solveHamiltonianCycle_DegreeOrdering(vector<vector<int>>& graph) {
    int n = graph.size();
    vector<int> path(n, -1);
    vector<bool> visited(n, false);
    path[0] = 0;
    visited[0] = true;

    // Calculate the degree of each vertex.
    vector<int> degree(n, 0);
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (graph[i][j] == 1)
                degree[i]++;
        }
    }

    if (findHamiltonianCycle_DegreeOrdering(graph, path, 1, 0, visited, degree)) {
        cout << "Hamiltonian Cycle found (Degree Ordering): ";
        for (int v : path)
            cout << v << " ";
        cout << "0" << endl;
    } else {
        cout << "No Hamiltonian Cycle exists (Degree Ordering)" << endl;
    }
}

// 4. Backtracking with Warnsdorff's Rule (for Knight's Tour on a chessboard)
//    - A heuristic specifically designed for finding Knight's Tours (a special case of Hamiltonian paths).
//    - Prioritizes moving to the square with the fewest possible next moves.
//    -  This approach is specific to Knight's tour, but I'll adapt it to work (less efficiently) for a general graph to demonstrate the concept.  It's not ideal for general Hamiltonian cycles.
//    - Time Complexity:  Difficult to express generally, but often very efficient for Knight's Tours.  Less so for general graphs.
//    - Space Complexity: O(n)
bool findHamiltonianCycle_Warnsdorff(vector<vector<int>>& graph, vector<int>& path, int pos, int start, vector<bool>& visited) {
    int n = graph.size();

    if (pos == n) {
        if (graph[path[pos - 1]][start] == 1)
            return true;
        else
            return false;
    }

    vector<pair<int, int>> candidates; // (vertex, number of unvisited neighbors)
      for (int v = 0; v < n; v++) {
        if (isSafe(v, path, graph, pos) && !visited[v]) {
            int unvisitedNeighbors = 0;
            for (int neighbor = 0; neighbor < n; neighbor++) {
                if (graph[v][neighbor] == 1 && !visited[neighbor]) {
                    unvisitedNeighbors++;
                }
            }
            candidates.push_back({v, unvisitedNeighbors});
        }
    }

    // Sort candidates based on the number of unvisited neighbors (Warnsdorff's rule).
    sort(candidates.begin(), candidates.end(), [](const auto& a, const auto& b) {
        return a.second < b.second;
    });

    for (const auto& candidate : candidates) {
        int v = candidate.first;
        path[pos] = v;
        visited[v] = true;
        if (findHamiltonianCycle_Warnsdorff(graph, path, pos + 1, start, visited))
            return true;
        path[pos] = -1;
        visited[v] = false;
    }
    return false;
}

void solveHamiltonianCycle_Warnsdorff(vector<vector<int>>& graph) {
    int n = graph.size();
    vector<int> path(n, -1);
    vector<bool> visited(n, false);
    path[0] = 0;
    visited[0] = true;

    if (findHamiltonianCycle_Warnsdorff(graph, path, 1, 0, visited)) {
        cout << "Hamiltonian Cycle found (Warnsdorff's): ";
        for (int v : path)
            cout << v << " ";
        cout << "0" << endl;
    } else {
        cout << "No Hamiltonian Cycle exists (Warnsdorff's)" << endl;
    }
}

// 5. Iterative Backtracking (Stack-based)
//    - Implements backtracking using a stack instead of recursion.
//    - Can sometimes be more efficient than recursion due to reduced overhead.
//    - Time Complexity: O(n!)
//    - Space Complexity: O(n)
void solveHamiltonianCycle_Iterative(vector<vector<int>>& graph) {
    int n = graph.size();
    vector<int> path(n, -1);
    vector<bool> visited(n, false);
    stack<pair<int, int>> s; // (vertex, position in path)
    int start = 0;

    path[0] = start;
    visited[start] = true;
    s.push({start, 1});

    while (!s.empty()) {
        int currentVertex = s.top().first;
        int pos = s.top().second;
        s.pop();

        if (pos == n) {
            if (graph[currentVertex][start] == 1) {
                cout << "Hamiltonian Cycle found (Iterative): ";
                for (int v = 0; v < n; v++)
                    cout << path[v] << " ";
                cout << start << endl;
                return;
            }
            else
                continue; // Important: Keep searching if the last vertex doesn't connect to start.
        }

        for (int v = 0; v < n; v++) {
            if (isSafe(v, path, graph, pos) && !visited[v]) {
                path[pos] = v;
                visited[v] = true;
                s.push({v, pos + 1});
                break; // Move to the next vertex
            }
             if (v == n - 1) {
                // Backtrack: If no safe vertex is found, reset and backtrack.
                path[pos-1] = -1;
                visited[currentVertex] = false;
             }
        }
    }
    cout << "No Hamiltonian Cycle exists (Iterative)" << endl;
}



int main() {
    // Example graph represented as an adjacency matrix.
    //  0 1 2 3
    //0 0 1 0 1
    //1 1 0 1 1
    //2 0 1 0 1
    //3 1 1 1 0
    vector<vector<int>> graph = {
        {0, 1, 0, 1},
        {1, 0, 1, 1},
        {0, 1, 0, 1},
        {1, 1, 1, 0}
    };
    // Expected output (one possible solution):
    // Hamiltonian Cycle found: 0 1 2 3 0

    cout << "Graph:" << endl;
    for(const auto& row : graph){
        for(int val : row){
            cout << val << " ";
        }
        cout << endl;
    }
    cout << endl;

    solveHamiltonianCycle_Recursive(graph);
    solveHamiltonianCycle_ForwardChecking(graph);
    solveHamiltonianCycle_DegreeOrdering(graph);
    solveHamiltonianCycle_Warnsdorff(graph); // Note: Not ideal for general graphs.
    solveHamiltonianCycle_Iterative(graph);

    return 0;
}
